p2p/host/basic/basic_host_notifiee.go

```diff
@@ -0,0 +1,41 @@
+package basichost
+
+import (
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// basicHostNotifiee listens to Network Notifiee events & emits them on the host's event bus
+type basicHostNotifiee BasicHost
+
+func (b *basicHostNotifiee) basicHost() *BasicHost {
+	return (*BasicHost)(b)
+}
+
+func (b *basicHostNotifiee) Connected(n network.Network, v network.Conn) {
+	// Is this our first connection with the peer ?
+	p := v.RemotePeer()
+	if len(n.ConnsToPeer(p)) == 1 {
```

Comment:
`n.Connectedness` should be faster.

We also need to take some kind of lock (ideally striped per peer?). Otherwise, we could have the following ordering:

* Thread 1: Receive disconnect event.
* Thread 1: Determine that we are disconnected.
* Thread 2: Receive connect event.
* Thread 2: Determine that we have a connection.
* Thread 2: Fire connect event.
* Thread 1: Fire disconnect event.

This is why we take locks in notification handlers in bitswap, the dht, etc. This is _also_ why I'm excited to finally have this event bus. We can solve this issue once _here_ and downstream callers can know that events will be completely ordered.

Reply:
@Stebalien I don't think this problem can happen here because both the `BasicHost.Connected` & `BasicHost.Disconnected` callbacks are writing to the same emitter.

1. Swarm takes a lock before calling the callbacks for an event & releases it ONLY after all callbacks have returned.

```
var wg sync.WaitGroup

	s.notifs.RLock()
	wg.Add(len(s.notifs.m))
	for f := range s.notifs.m {
		go func(f network.Notifiee) {
			defer wg.Done()
			notify(f)
		}(f)
	}

	wg.Wait()
	s.notifs.RUnlock()
```

2. For the scenario that you highlight, the `BasicHost.Disconnected` callback will return ONLY after the event has been written to the `EvePeerStateChange` emitter. So, can't we be assured that the subsequent Connect will be written to the same emitter ONLY after the disconnect event has been written & this will be received in order by the subscriber ? 

I might be missing something here. Let me know.

Reply:
>     Swarm takes a lock before calling the callbacks for an event & releases it ONLY after all callbacks have returned.

That's a read lock.

---

Note: for a single connection, the disconnect event will always come after the connect event. However, for separate connections to the same peer, there are no ordering guarantees.

Reply:
I got what you mean. Making the change.

Reply:
@Stebalien I am not using `n.Connectedness ` as we ONLY want to fire an event if this is our first connection with the peer.

Reply:
> I am not using n.Connectedness as we ONLY want to fire an event if this is our first connection with the peer.

Ah, I didn't see that. Unfortunately, this is unreliable as we might create two connections before firing the events.

Reply:
This is a way trickier beast than I'd imagined.